tester/src/test/scala/spinal/tester/scalatest/LatchTester.scala

```diff
@@ -4,36 +4,99 @@ import org.scalatest.funsuite.AnyFunSuite
 import spinal.core._
 import spinal.core.sim._
 
+import scala.util.Random
+
 class LatchTester extends AnyFunSuite {
   test("bad latch") {
-    SimConfig.compile(new Component {
-      val myCond   = Bool
-      val inVal    = Bits(8 bits)
-      val latchVal = Bits(8 bits)
+    var didRaise = false
 
-      when(myCond) {
-        latchVal := inVal
-      }
-    })
+    try {
+      SimConfig.compile(new Component {
+        val myCond = Bool
+        val inVal = Bits(8 bits)
+        val latchVal = Bits(8 bits)
+
+        when(myCond) {
+          latchVal := inVal
+        }
+      })
+    } catch {
+      case e: SpinalExit if e.getMessage().contains("LATCH DETECTED") =>
+        didRaise = true
+    }
+
+    assert(didRaise)
   }
 
   test("allowed latch") {
     SimConfig.compile(new Component {
-      val myCond   = Bool
-      val inVal    = Bits(8 bits)
+      definition.addComment("verilator lint_off LATCH")
+
+      val io = new Bundle {
+        val myCond   = in Bool()
+        val inVal    = in Bits(8 bits)
+        val outVal   = out Bits(8 bits)
+      }
+
       val latchVal = Latch(Bits(8 bits))
 
-      when(myCond) {
-        latchVal := inVal
+      when(io.myCond) {
+        latchVal := io.inVal
       }
+
+      io.outVal := latchVal
+    }).doSim(dut => {
+      dut.io.inVal  #= Random.nextInt(255)
+      dut.io.myCond #= false
+      sleep(10)
+
+      // Trigger the latch
+      dut.io.myCond #= true
+      sleep(1)
+      dut.io.myCond #= false
+
+      // Check that the latch took the value
+      assert(dut.io.inVal.toInt == dut.io.outVal.toInt)
+      dut.io.inVal  #= Random.nextInt(255)
+
+      // Wait a bit
+      sleep(10)
+
+      // Check that the latch did not take the value
+      assert(dut.io.inVal.toInt != dut.io.outVal.toInt)
     })
   }
 
   test("latchWhen") {
     SimConfig.compile(new Component {
-      val myCond   = Bool
-      val inVal    = Bits(8 bits)
-      val latchVal = LatchWhen(inVal, myCond)
+      definition.addComment("verilator lint_off LATCH")
+
+      val io = new Bundle {
+        val myCond = in Bool()
+        val inVal = in Bits (8 bits)
+        val outVal = out Bits (8 bits)
+      }
+      io.outVal := LatchWhen(io.inVal, io.myCond)
+    }).doSim(dut => {
```

Comment:
Idea to avoid test code duplication.

Here we are testing two different configurations of an entity, in VHDL words.

(This is draft code written in GitHub comments, did not test it)

```scala
// In LatchTester
class Dut extends Component {
  val io = new Bundle {
    val cond = in Bool ()
    val input = in Bits (8 bits)
    val output = out Bits (8 bits)
  }
}

class DutWhen extends Dut {
  val latch = Latch(Bits(8 bits))
  when (io.cond) { latch := io.input }
  io.output := latch
}

class DutLatchWhen extends Dut {
  io.output := LatchWhen(io.input, io.cond)
}

def protocol(dut: Dut): Unit = {
  // the simulation code here
}

test("Latch") {
  SimConfig.compile(DutConfigWhen).doSim(protocol)
}

test("LatchWhen") {
  SimConfig.compile(DutConfigLatchWhen).doSim(protocol)
}
```

Note: could this go in https://github.com/SpinalHDL/SpinalDoc-RTD/issues/131?

Reply:
That largely worked as you expected it to. The only hang up I had was with the `.doSim(protocol)`. I had to do either `.doSim(protocol _)` or `.doSim(protocol(_))`. You'll see it in the commit I made. (I also found `simDriver` to be a more descriptive name for the method)

Thanks for the suggestion! I knew there had to be a way to keep it DRY

Reply:
Ah yes, maybe defining the protocol as below would have allowed the `.doSim(protocol)`. But that's not important. Thanks again for your work!

```scala
val protocol: Dut => Unit = { dut =>
  // the simulation code here
}
```